File: src/main/java/ru/kainlight/lightshowregion/COMMON/lightlibrary/LightPlugin.java
package ru.kainlight.lightshowregion.COMMON.lightlibrary;

import lombok.Getter;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;
import ru.kainlight.lightshowregion.COMMON.lightlibrary.CONFIGS.BukkitConfig;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

@Getter
public class LightPlugin extends JavaPlugin {

    private final double CONFIG_VERSION = 1.0;
    private final boolean paper = isPaperPlugin();
    public BukkitConfig messageConfig;

    @Override
    public void onLoad() {
        this.saveDefaultConfig();
        BukkitConfig.saveLanguages(this, "language");
        updateConfig();
        messageConfig.updateConfig();
    }

    @SuppressWarnings("all")
    protected void updateConfig() {
        // Загрузка текущей конфигурации
        FileConfiguration userConfig = getConfig();
        double version = userConfig.getDouble("config-version");
        if(version == CONFIG_VERSION) return;

        // Чтение конфигурации по умолчанию из JAR-файла
        InputStream defaultConfigStream = getResource("config.yml");
        InputStreamReader inputConfigReader = new InputStreamReader(defaultConfigStream, StandardCharsets.UTF_8);
        YamlConfiguration defaultConfig = YamlConfiguration.loadConfiguration(inputConfigReader);

        // Добавление отсутствующих значений из конфигурации по умолчанию и удаление удаленных ключей
        defaultConfig.getKeys(true).forEach(key -> {
            // Если у пользователя нет такого ключа или его значение является значением по умолчанию, добавляем его
            if (!userConfig.contains(key) || userConfig.get(key).equals(defaultConfig.get(key))) {
                userConfig.set(key, defaultConfig.get(key));
            }
        });

        userConfig.getKeys(true).forEach(key -> {
            if (!defaultConfig.contains(key)) {
                userConfig.set(key, null);
            }
        });

        getLogger().warning("config.yml updated");
        getConfig().set("config-version", CONFIG_VERSION);
        saveConfig();
    }

    private static boolean isPaperPlugin() {
        try {
            Class.forName("com.destroystokyo.paper.ParticleBuilder");
            return true;
        } catch (ClassNotFoundException e) {
            return false;
        }
    }

}

File: src/main/java/ru/kainlight/lightshowregion/library/LightPlugin.kt
package ru.kainlight.lightshowregion.library

import org.bukkit.Bukkit
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

open class LightPlugin : JavaPlugin() {

    private var configurationVersion = 1.2
    val isPaper: Boolean = isPaperPlugin()

    // -------------------- //
    fun registerListener(listener: Listener): LightPlugin {
        Bukkit.getPluginManager().registerEvents(listener, this)
        return this
    }

    fun registerCommand(command: String, executor: CommandExecutor, tabCompleter: TabCompleter? = null): LightPlugin {
            this.getCommand(command)?.let {
            it.setExecutor(executor)
            if (tabCompleter != null) it.tabCompleter = tabCompleter
        }
        return this
    }

    // ------------------------ //
    fun updateConfig() {
        // Загрузка текущей конфигурации
        val userConfig = config
        val version = userConfig.getDouble("config-version")
        if (version == configurationVersion) return

        // Чтение конфигурации по умолчанию из JAR-файла
        val defaultConfigStream = getResource("config.yml") ?: return
        val inputConfigReader = InputStreamReader(defaultConfigStream, StandardCharsets.UTF_8)
        val defaultConfig = YamlConfiguration.loadConfiguration(inputConfigReader)

        // Добавление отсутствующих значений из конфигурации по умолчанию и удаление удаленных ключей
        defaultConfig.getKeys(true).forEach { key: String ->
            // Если у пользователя нет такого ключа или его значение является значением по умолчанию, добавляем его
            if (!userConfig.contains(key) || userConfig[key] == defaultConfig[key]) {
                userConfig[key] = defaultConfig[key]
            }
        }

        userConfig.getKeys(true).forEach { key: String ->
            if (!defaultConfig.contains(key)) {
                userConfig[key] = null
            }
        }

        /// logger.warning("config.yml updated")
        config["config-version"] = configurationVersion
        saveConfig()
    }

    private fun isPaperPlugin(): Boolean {
        try {
            Class.forName("com.destroystokyo.paper.ParticleBuilder")
            return true
        } catch (e: ClassNotFoundException) { return false }
    }

    fun command(command: String, sender: CommandSender = Bukkit.getConsoleSender()): Boolean {
        return Bukkit.dispatchCommand(sender, command)
    }
}